flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/StopWithSavepointTerminationHandlerImpl.java

```
@@ -0,0 +1,258 @@
+ */
+public class StopWithSavepointTerminationHandlerImpl
+        implements StopWithSavepointTerminationHandler {
+
+    private final Logger log;
+
+    private final SchedulerNG scheduler;
+    private final CheckpointScheduling checkpointScheduling;
+    private final JobID jobId;
+
+    private final CompletableFuture<String> result = new CompletableFuture<>();
+
+    private State state = new WaitingForSavepoint();
+
+    public <S extends SchedulerNG & CheckpointScheduling> StopWithSavepointTerminationHandlerImpl(
+            @Nonnull JobID jobId, @Nonnull S schedulerWithCheckpointing, @Nonnull Logger log) {
+        this(jobId, schedulerWithCheckpointing, schedulerWithCheckpointing, log);
+    }
+
+    @VisibleForTesting
+    StopWithSavepointTerminationHandlerImpl(
+            @Nonnull JobID jobId,
+            @Nonnull SchedulerNG scheduler,
+            @Nonnull CheckpointScheduling checkpointScheduling,
+            @Nonnull Logger log) {
+        this.jobId = checkNotNull(jobId);
+        this.scheduler = checkNotNull(scheduler);
+        this.checkpointScheduling = checkNotNull(checkpointScheduling);
+        this.log = checkNotNull(log);
+    }
+
+    @Override
+    public CompletableFuture<String> handlesStopWithSavepointTermination(
+            CompletableFuture<CompletedCheckpoint> completedSavepointFuture,
+            CompletableFuture<Collection<ExecutionState>> terminatedExecutionsFuture,
+            ComponentMainThreadExecutor mainThreadExecutor) {
+        completedSavepointFuture
+                .whenCompleteAsync(
+                        (completedSavepoint, throwable) -> {
+                            if (throwable != null) {
+                                handleSavepointCreationFailure(throwable);
+                            } else {
+                                handleSavepointCreation(completedSavepoint);
+                            }
+                        },
+                        mainThreadExecutor)
```

I would suggest to consume an exception via `handleAsync` and then add a `ExceptionUtils.assertNoException` in order to make potential misusages of the state call visible. Otherwise if `handleSavepoint*` fails, then this exception will be swallowed by the returned future.